### ISSUE TITLE
BUG: assert_equal does not distinguish scalars from 0-dim arrays

### ISSUE THREAD

**Issue by h-vetinari** [numpy/numpy#24050]
### Describe the issue:

I suspect this is part of the larger issue of consistency (resp. lack thereof) around scalars vs. 0-dim arrays.

But it's worse in this case, because now any user of the various `assert_*` functions from `numpy.testing` will be themselves unable to ensure that the behaviour is as intended.

### Reproduce the code example:

```python
>>> import numpy as np
>>> from numpy.testing import assert_equal
>>> assert_equal(1.0, np.array(1, dtype=np.int64))  # passes?!?
>>> assert_equal(1.0, np.array(2, dtype=np.int64))  # changed value -> fails
Traceback (most recent call last):
[...]
```

### Runtime information:

```
>>> import sys, numpy; print(numpy.__version__); print(sys.version)
1.25.0
3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:34:57) [MSC v.1936 64 bit (AMD64)]
```

```
>>> print(numpy.show_runtime())
WARNING: `threadpoolctl` not found in system! Install it by `pip install threadpoolctl`. Once installed, try `np.show_runtime` again for more detailed build information
[{'numpy_version': '1.25.0',
  'python': '3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:34:57) '
            '[MSC v.1936 64 bit (AMD64)]',
  'uname': uname_result(system='Windows', node='AXEL-X1', release='10', version='10.0.22621', machine='AMD64')},
 {'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2'],
                      'not_found': ['AVX512F',
                                    'AVX512CD',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL']}}]
None
```

### Context for the issue:

`numpy.testing` is a key component for ensuring people can build their own libraries on top of numpy. "equal" should actually mean equal type, shape, content, etc. (as it usually does...), also for 0-dim arrays.

**Comment by h-vetinari** [comment]
Workaround is now:
```python
        # otherwise scalars and 0-dim arrays of the same value compare equal
        def assert_really_equal(x, y):
            assert type(x) == type(y), f"types not equal: {type(x)}, {type(y)}"
            assert_equal(x, y)
```

Needless to say, this feels a bit ridiculous.

**Comment by WarrenWeckesser** [comment]
FYI: In NumPy 1.24.0, the `strict` parameter [was added to](https://github.com/numpy/numpy/pull/21595) [`assert_array_equal`](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_array_equal.html), but it will still consider a Python or NumPy scalar equal to an array scalar of size 0, e.g. `assert_array_equal(np.int64(0), np.array(0, dtype=np.int64), strict=True)` will pass, because they have the same shape and dtype.  Perhaps it should be even stricter, and fail in that case?

**Comment by seberg** [comment]
My guess is that there are a giant amount of tests which do `assert_equal(arr, [1, 2, 3])` including in SciPy.

The strict parameter could be even stricter, I have no opinion on whether it should check types; maybe types are just meant to be checked explicitly.

**Comment by h-vetinari** [comment]
I think what's confusing is that other `assert_*` can turn out quite strict in many other aspects, which is why there's `assert_allclose` etc., which go a bit less hard. I remember[^1] that failed on different types (yet my example above compares 1.0 & 1 successfully). That makes it seem even more of a surprise if big differences in (container) types are suddenly ignored.

[^1]: though that wasn't `assert_equal` I guess; I just tried and fail to reproduce that

I don't have strong opinions on how this should be implemented, but I guess a `strict=` keyword would make sense (especially as it could be similarly added to other methods; i.e. it'd be well-defined what `assert_allclose(..., strict=True)` would mean: up to a defined tolerance, but no type variance).

I also think that making the existing `strict=True` even stricter doesn't even count as "breakage", because whoever's using that from 1.24 is _already_ explicitly opting into "the default is too loose for me".

**Comment by h-vetinari** [comment]
> I just want to point out that this is true:

Absolutely, because the requirements of what most people need for processing their _data_  is a by-value comparison.

However, for testing (which `numpy.testing.assert_*` fall under quite clearly), you often need to check more tightly, because it's necessary - for example - to maintain other invariants (like using the array API on the result). My argument is: if numpy doesn't offer a strict(er) version, then it becomes really hard to test _at all_.

**Comment by seberg** [comment]
There are two things to keep in mind:
* Sentiment about type strictness increased a lot over time and is also not universally agreed upon (the opposite bug report wouldn't surprise me if changed).
* I suspect that NumPy and SciPy each would have hundreds of test failures alone if this was changed.  (Yes, they are easy, but is it worth it?)

We discussed `strict=` addition and tightening that kwarg which I suspect we would get consensus on.  Maybe you can suggest other new API also?
But attempting a noisy transition seems like asking too much from other projects.
